### PR TITLE
feat: telemetry push

### PR DESCRIPTION
**Pull request by Ewan-Roberts**
## Changes

- Added ability to send telemetry data to WCP endpoint
- Added telemetry code to wrap clients lambda functions
- Added telemetry code pulling for getting latest telemetry code from remote bucket

AWAITING:
- Clients WCP_API_KEY is to be set

## How Has This Been Tested?

- Jest

## Documentation

TBC this should possibly be added to legal documentation or security pages

### CODE REVIEW

File: packages/project-utils/package.json [hunk 1]
```diff
@@ -41,6 +42,7 @@
     "get-yarn-workspaces": "1.0.2",
     "glob": "^7.1.2",
     "html-webpack-plugin": "4.0.0-beta.5",
+    "https": "^1.0.0",
```
Comment (adrians5j) [inline packages/project-utils/package.json]: Hm... why `https` here?

File: packages/project-utils/package.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   "name": "@webiny/project-utils",
+  "telemetryVersion": "1.0.1",
```
Comment (adrians5j) [inline packages/project-utils/package.json]: Can you describe this a bit to me?

File: package.json [hunk 3]
```diff
@@ -228,5 +228,9 @@
     "@pulumi/pulumi": "3.19.0",
     "@pulumi/aws": "4.28.0"
   },
-  "packageManager": "yarn@3.1.0"
+  "packageManager": "yarn@3.1.0",
```
Comment (adrians5j) [inline package.json]: 1. These new dependencies should not be defined here, but in the actual package / `package.json` that is using them. 2. `https` is a native Node module, doesn't need to be included in list of dependencies.

File: packages/project-utils/package.json (hunk 1 above)
Comment (adrians5j) [reply packages/project-utils/package.json]: BTW as mentioned in other comment, native Node modules do not need to be declared as dependencies. Feel free to remove this.

File: api/code/graphql/package.json [hunk 4]
```diff
@@ -47,5 +47,8 @@
         "@webiny/api-tenant-manager"
       ]
     }
+  },
+  "devDependencies": {
+    "graphql-request": "^3.4.0"
```
Comment (adrians5j) [inline api/code/graphql/package.json]: Hmm.. why is this here?

File: packages/project-utils/bundling/function/buildFunction.js [hunk 5]
```diff
@@ -54,4 +73,21 @@ module.exports = options => {
             resolve();
         });
     });
+
+    const handlerFile = fs.readFileSync(options.cwd + "/build/handler.js", {
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: Use `path.join` to join paths. This is mainly for compatibility with multiple OSes.

File: packages/project-utils/bundling/function/buildFunction.js [hunk 6]
```diff
@@ -54,4 +73,21 @@ module.exports = options => {
...
+        encoding: "utf8",
+        flag: "r"
+    });
+    const includesGraphQl = handlerFile.includes("handler-graphql");
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: How does this perform BTW? Is it an intensive operation / an operation that takes a lot of time to execute? Just wondering...

File: packages/project-utils/bundling/function/buildFunction.js [hunk 7]
```diff
@@ -54,4 +73,21 @@ module.exports = options => {
...
+
+    if (includesGraphQl) {
+        try {
+            await getLatestTelemetryCode;
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: You mean `await getLatestTelemetryCode();` ?

Comment (adrians5j) [reply packages/project-utils/bundling/function/buildFunction.js]: Let's rename this function. `get` prefix in function name usually indicates that the function is returning a value, but in this case, the function is really just adapting the manipulating built code.

Let's call it for example `injectHandlerTelemetry`.

File: packages/project-utils/bundling/function/buildFunction.js [hunk 8]
```diff
@@ -54,4 +73,21 @@ module.exports = options => {
...
+        } catch (err) {
+            console.log("Error loading telemtry code");
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: These should throw an error, and not do `console.log`.

**Review by adrians5j — CHANGES_REQUESTED** [review]
First iteration.

Let's check these couple of quick fixes first. BTW then `package-lock.json` should not be in the repo. We use `yarn`, which relies on `yarn.lock`.

File: packages/project-utils/package.json (hunk 2 above)
Comment (Ewan-Roberts) [reply packages/project-utils/package.json]: So I wanted a place to store the local telemetry version number so I can compare it to the remote version so thought it best to store it in the `package.json`. Does that make sense or were you thinking of storing it somewhere else?

File: packages/project-utils/bundling/function/buildFunction.js (hunk 6 above)
Comment (Ewan-Roberts) [reply packages/project-utils/bundling/function/buildFunction.js]: this bundle was on the smaller size but its pretty quick at 10ms on average

File: package.json (hunk 3 above)
Comment (Ewan-Roberts) [reply package.json]: so I added this due to an error thrown by adio: 
```
✖ "adio" found some errors. Please fix them and try committing again.
1. @webiny/project-utils (/Users/ewan/Development/webiny/webiny-js/packages/project-utils)
Packages used in source code, but not listed in package.json:
1. https
```
I wonder why adio is picking this up as a required library 🤔 
I have kept it in so i can commit and push for the time being

File: packages/project-utils/package.json (hunk 1 above)
Comment (Ewan-Roberts) [reply packages/project-utils/package.json]: same as https://github.com/webiny/webiny-js/pull/2128#discussion_r776326749

File: packages/project-utils/bundling/function/telemetry.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    const latestTelemetryVersion = telemetryVersions.pop();
+
+    if (packageData.telemetryVersion === latestTelemetryVersion) {
+        return;
+    }
+
+    const latestCodeData = await s3
+        .getObject({
+            Bucket: "telemetry-code",
+            Key: "telemetry-versions/" + latestVersion
```
Comment (Ewan-Roberts) [inline packages/project-utils/bundling/function/telemetry.js]: hmmmm maybe a dumb qustion but will this work on another cilents instance? Its a public bucket, but bucket names are not unique, so how does aws know where to get it from?

File: packages/project-utils/package.json (hunk 1 above)
Comment (adrians5j) [reply packages/project-utils/package.json]: Ah...

Yeah, just add it here `.adiorc.js:14`. That should then let you commit without setting `https` in deps.

File: packages/project-utils/package.json (hunk 2 above)
Comment (adrians5j) [reply packages/project-utils/package.json]: Hm.. maybe I'm missing something.. would it work if we just relied on `@webiny/project-utils`'s version? So on the backend, we just take that version into account and apply the needed logic accordingly?

BTW on backend, there is the `process.env.WEBINY_VERSION` env var, or even better: `context.WEBINY_VERSION` (`packages/handler/src/types.ts`).

Maybe we can discuss this further.

But in any case, let's write a short document in Notion / WCP category on this topic. 

Something like - what are the problems we're trying to achieve this. What happens in the client is not up-to-date. And so on.

File: packages/project-utils/bundling/function/buildFunction.js (hunk 6 above)
Comment (adrians5j) [reply packages/project-utils/bundling/function/buildFunction.js]: Nice :)

I just wanted to confirm it's not like 30secs or something like that :)

File: packages/project-utils/bundling/function/buildFunction.js [hunk 10]
```diff
@@ -54,4 +73,20 @@ module.exports = options => {
             resolve();
         });
     });
+
+    const handlerFile = fs.readFileSync(path.join(options.cwd, "/build/handler.js"), {
+        encoding: "utf8",
+        flag: "r"
+    });
+    const includesGraphQl = handlerFile.includes("handler-graphql");
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: We'll need to include `handler-graphql-content-model` (`packages/api-headless-cms/src/content/graphQLHandlerFactory.ts:127`) too, because this is the plugin name of the dynamic Headless CMS GraphQL API (so, the `handler-graphql` package is not being used for that).

File: packages/project-utils/bundling/function/buildFunction.js [hunk 11]
```diff
@@ -54,4 +73,20 @@ module.exports = options => {
...
+
+    if (includesGraphQl) {
+        try {
+            await injectHandlerTelemetry();
+        } catch (err) {
+            throw new Error(err);
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: I guess we can remove the `catch (err)` since the only thing we're doing within the catch block is again throwing the same `err`. :)

File: packages/project-utils/bundling/function/buildFunction.js [hunk 12]
```diff
@@ -54,4 +73,20 @@ module.exports = options => {
             resolve();
         });
     });
+
+    const handlerFile = fs.readFileSync(path.join(options.cwd, "/build/handler.js"), {
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: I just noticed we'll need to add a check here, before performing this handler code-related stuff.

Basically, if the `id` of the project is not present in root `webiny.project.ts` file, the whole dev/deploy experience should work as if the user is not using WCP at all.

This is how things work in my yet-to-be-merged branch.

So, ultimately, what we need to do here is do a check - if `!project.config.id`, then exit.

The `project` is something you can get by calling `getProject`, for example: `packages/cli-plugin-scaffold-react-component/src/index.ts:117`. Note that you'll need to pass the `cwd` from `options` object to it.

File: packages/project-utils/bundling/function/buildFunction.js [hunk 13]
```diff
@@ -1,8 +1,27 @@
...
+const fs = require("fs");
+const telemetry = require("./telemetry");
 
-module.exports = options => {
+async function injectHandlerTelemetry() {
+    await telemetry.getLatestTelemetryFunction();
+
+    fs.copyFileSync(
+        path.join(cwd, "build", "handler.js"),
+        path.join(cwd, "build", "handler.original.js")
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: Let's rename `handler.original.js` to `_handler.js`. It's nothing super important, but my thinking is maybe the less the user knows here, the better.

**Review by adrians5j — CHANGES_REQUESTED** [review]
I left a couple of comments. But you know what.

I think for now I'm gonna pause the review, so that we can, as mentioned in one of the comments below, first create a short document that describes how telemetry client will be retrieved on deployments, and how it works.

Let's do that first, and then we'll see if some more important changes will be needed, and if this code (mainly `packages/project-utils/bundling/function/telemetry.js`) will need to change.

File: packages/project-utils/bundling/function/buildFunction.js (hunk 10 above)
Comment (Ewan-Roberts) [reply packages/project-utils/bundling/function/buildFunction.js]: Maybe I miss understand, this will capture anything with the string "handler-graphql" so this should be included no?

File: packages/app-page-builder/src/editor/plugins/elementSettings/align/HorizontalAlignFlexSettings.tsx [hunk 14]
```diff
@@ -67,78 +67,77 @@ enum AlignmentsTypeEnum {
...
-        const element = useRecoilValue(elementWithChildrenByIdSelector(activeElementId));
-        const fallbackValue = useMemo(
-            () =>
-                applyFallbackDisplayMode(displayMode, mode =>
-                    get(element, `${DATA_NAMESPACE}.${mode}`)
-                ),
-            [displayMode]
-        );
-        const align = get(element, propName, fallbackValue || AlignmentsTypeEnum.CENTER);
+const HorizontalAlignFlexSettings: React.FunctionComponent<
```
Comment (Ewan-Roberts) [inline packages/app-page-builder/src/editor/plugins/elementSettings/align/HorizontalAlignFlexSettings.tsx]: Anything under app-page-builder can be ignored, they are prettier fixes

File: packages/project-utils/bundling/function/telemetry.js [hunk 15]
```diff
@@ -0,0 +1,16 @@
+const fs = require("fs");
+const fetch = require("node-fetch");
+
+const TELEMETRY_BUCKET_URL = "https://wcp-telemetry-function-7cbe312.s3.amazonaws.com/latest";
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/telemetry.js]: Hm.. let's do this in the following way...

In the WCP repo where you're doing the hosting of the client code... let's rather have a CF distribution that is configured to serve files from S3 when `/clients/*` is requested. The files are ofc pulled from S3.

So, this URL will always be something like `xyz.cloudfront.net/clients/latest`. Something along those lines.

Also, please introduce an option to override this URL via an env variable, so that we can do some testing (automated or manual) in the future here.

I know for example I added the following two so that I can perform some testing with Webiny CLI stuff:

```bash
WCP_API_URL=https://d2d4evj0p7twjc.cloudfront.net/graphql
WCP_APP_URL=http://localhost:3000
```

So.. maybe we could also have like `WCP_API_CLIENTS_URL` or something like that.

File: packages/project-utils/bundling/function/telemetryFunction.js [hunk 16]
```diff
@@ -0,0 +1,116 @@
+const userFunction = require("./_handler.js");
+const https = require("https");
+const packageData = require("../../package.json");
+const AWS = require("aws-sdk");
+AWS.config.update({ region: "us-east-1" });
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/telemetryFunction.js]: Ughh we definitely don't want to globally override AWS SDK's region. What if a Lambda function was deployed into `eu-central-1`?  This could definitely affect the user if he's working with services within a different region.

Can we have an isolated instance of the AWS SDK, just so we don't affect the user?

File: packages/project-utils/bundling/function/buildFunction.js (hunk 10 above)
Comment (adrians5j) [reply packages/project-utils/bundling/function/buildFunction.js]: Nope, that's the thing (and a good question you've asked).

Interestingly enough, the Headless CMS Lambda function doesn't include the `handler-graphql` string. We'll see how to maybe better resolve this, by potentially not even relying on `handler-graphql`, but simply putting our own magic string into the code, which will be explicitly be used for these purposes. I'll write this into jira.

File: packages/project-utils/bundling/function/buildFunction.js [hunk 17]
```diff
@@ -1,8 +1,25 @@
...
 const chalk = require("chalk");
+const fs = require("fs");
+const telemetry = require("./telemetry");
+import { getProject } from "@webiny/cli/utils";
 
-module.exports = options => {
+async function injectHandlerTelemetry() {
+    await telemetry.updateTelemetryFunction();
+
+    fs.copyFileSync(path.join(cwd, "build", "handler.js"), path.join(cwd, "build", "_handler.js"));
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/buildFunction.js]: Wait, where is this `cwd` coming from? 🤔

**Review by adrians5j — DISMISSED** [review]
Let's continue by resolving some of the things reported in this iteration.

Thank you!🍻

File: packages/project-utils/bundling/function/telemetryFunction.js (hunk 16 above)
Comment (Ewan-Roberts) [reply packages/project-utils/bundling/function/telemetryFunction.js]: whoops this is an older version than the one in s3

File: packages/project-utils/bundling/function/telemetry.js (hunk 15 above)
Comment (Ewan-Roberts) [reply packages/project-utils/bundling/function/telemetry.js]: The CF distribution was very time consuming and had no progress even with help from the team, i have put it into a ticket so i can address it on your return: https://webiny.atlassian.net/browse/WEB-707

File: packages/project-utils/bundling/function/buildFunction.js (hunk 10 above)
Comment (Ewan-Roberts) [reply packages/project-utils/bundling/function/buildFunction.js]: For whoever reviews this, this is an outstanding question. I was thinking of adding a comment with `// add-function-to-telemetry-tracking` to the functions we want tracked and search for that string in the code

Comment (Pavel910) [reply packages/project-utils/bundling/function/buildFunction.js]: For the `alpha`, you can hardcode just the `graphql` and `headlessCMS` functions. As we're working on the new project setup, you will be able to hook into any project function via a plugin, so no string hardcoding will be necessary.

It's something new we're developing, so we'll have way more control over the generated code for the users (basically, there will be no code in the project at all, and we'll be generating stuff on fly). This means that you'll be able to intercept lambda templates, modify stuff with ASTs, or jsut add WCP plugins on the fly.

Comment (Pavel910) [reply packages/project-utils/bundling/function/buildFunction.js]: Resolved this over Slack and a call: to avoid problems with comments, which can be stripped away by webpack, let's register a dummy plugin, and use its `type` as your identifier which will tell you that you need to wrap the function with a tracker. Something like:

```js
{ type: "wcp-telemetry-tracker" }
```

My suggestions is to stick it in here: `packages/api-tenancy/src/index.ts:41`
This package is used by GraphQL, Headless CMS, and PB import/export functions. This will cover the WCP Alpha needs, and in the future, the new project setup will allow us to make the implementation a lot smarter than that.

File: packages/project-utils/bundling/function/telemetryFunction.js [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        return;
+    }
+
+    timerRunning = true;
+
+    setInterval(async () => {
+        const timeInFiveMinutes = Date.now() + MAXIMUM_MINUTES_UNTIL_REQUEST_FIRED * 60000;
+        if (timeInFiveMinutes > initialTime) {
+            if (localData.logs.length > 0) {
+                await postTelemetryData(localData);
```
Comment (Pavel910) [inline packages/project-utils/bundling/function/telemetryFunction.js:65]: if this `postTelemetryData` throws an error, will this break the original Lambda execution? 🤔

File: packages/project-utils/bundling/function/buildFunction.js [hunk 19]
```diff
@@ -60,4 +77,22 @@ module.exports = options => {
...
         });
     });
+
+    const project = getProject({ cwd });
+
+    if (!project.config.id) {
+        return result;
+    }
+
+    const handlerFile = fs.readFileSync(path.join(options.cwd, "/build/handler.js"), {
```
Comment (Pavel910) [inline packages/project-utils/bundling/function/buildFunction.js]: Bundles can be several MBs large, around 3-4MB. So this should not be synchronous. Use an asynchronous read, and feel free to use the `fs-extra` package for all FS related stuff, we already have it in the `project-utils` package, it exposes proper promisified functions.

The reason is, we're building ~14 functions, some of which are large, so reading 14 large files in a synchronous manner can make the terminal freeze for a few seconds.

**Review by Pavel910 — COMMENTED** [review]
Left some questions and comments @Ewan-Roberts.

File: packages/project-utils/bundling/function/telemetryFunction.js (hunk 18 above)
Comment (Ewan-Roberts) [reply packages/project-utils/bundling/function/telemetryFunction.js:65]: Good point, I wrapped this in a try catch
